added hie_bios_path_prefix attribute to haskell_repl

ylecornec opened the pull request:
I added an attribute named `hie_bios_path_prefix` to the `haskell_repl` rule to address issue #1432.

This makes it possible for hie-bios to output paths relative to a different folder than the workspace root.

In order to use absolute paths, it is possible to use a placeholder in the prefix,
and then replace it with a call to `pwd` in the user defined [.hie-bios script](https://rules-haskell.readthedocs.io/en/latest/haskell-use-cases.html#configuring-ide-integration-with-ghcide).

aherrmann reviewed (commented):
That looks good, thank you!

Could you add a test-case as well? We should test both cases: No prefix and some prefix. 

Perhaps an `sh_test` that has a `data` dependency on a `haskell_repl`'s hie-bios output would work.

ylecornec commented:
I added a test case which contains two `hs_repl` rules, one with the `hie_bios_path_prefix` option and one without which generate hie-bios files.

Then a `sh_test` rule check that the prefix appears in the first file, and that if we remove this prefix using sed, then the two files are equals.

Some notes : 
 - Since the test uses a shell script, I added the `dont_test_on_windows` tag.
 - I also added a newline at the end of the hie-bios files since sed on mac complained about it.
 - I left some debug print which can be useful in case the test fails later, is this good practice ?

aherrmann commented on tests/hie-bios/BUILD.bazel:
```diff
@@ -0,0 +1,55 @@
+        ":hie-bios-file",
+        ":hie-bios-file-no-prefix",
+    ],
+    tags = ["dont_test_on_windows"],
```
```suggestion
```
This shouldn't be necessary. rules_haskell requires msys2 on Windows anyway, so a bash is availabl. And Bazel can generate executables from `.sh` scripts on Windows if a bash is a available.

aherrmann replied:
Just saw the [Windows CI failure](https://github.com/tweag/rules_haskell/runs/2325545499#step:6:102):
```
grep: tests/hie-bios/hie-bios@hie-bios: No such file or directory
```
I think the issue has to do with runfiles. On Unix Bazel will create a tree of symlinks in the tests working directory pointing to `data` dependencies. However, on Windows symlinks are not generally available, so instead Bazel generates a runfiles manifest file and code needs to use a runfiles library to query the manifest file for the location of `data` dependencies. The bash runfiles library is documented [here](https://github.com/bazelbuild/bazel/blob/master/tools/bash/runfiles/runfiles.bash#L39-L68). Example uses can be found in the repository. In tests you can use the `TEST_WORKSPACE` env-var to find the workspace name, e.g.
```
file1="$(rlocation "$TEST_WORKSPACE/$1")"
```

ylecornec replied:
Ah I see, thank you.

So I activated the windows test back on and it works with the bash runfiles library.
And now, debug information is only printed in case the test fails.

aherrmann reviewed (commented):
> I left some debug print which can be useful in case the test fails later, is this good practice ?

I think it's better to not print this output in the success case and only print it in the failure case.

---

It looks like there's a conflict with `master` in `haskell/repl.bzl`. Could you rebase on `master`?

aherrmann reviewed (approved):
Thank you, looks good!
